### PR TITLE
Add Windows/macOS/Linux and x86/arm64 fuzzing CI

### LINKED ISSUE #252: fuzz tests should be run from CI
As mentioned in #250 , the fuzz tests really should be run from CI.

### PR DESCRIPTION
**Pull request by avsm**
Azure Pipelines and cloud.drone.io

For the fuzz tests, this fixes #252, but how do we tell `afl-fuzz` how long to run for?  (See .drone.yml, tweaks welcome to the command line). It's running atm on https://cloud.drone.io/mirage/ocaml-cstruct/6/1/2 which I hope is a public URL.

### CODE REVIEW

**Comment by avsm** [comment]
Windows is broken due to https://github.com/janestreet/jst-config/issues/1 making cstruct-async uninstallable

**Comment by yomimono** [comment]
> how do we tell afl-fuzz how long to run for

I just used `timeout` for this, FWIW.

File: .drone.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  - sudo apt-get update && sudo apt-get -y install afl
+  - sudo chown -R opam .
+  - git -C /home/opam/opam-repository pull origin && opam update
+  - opam switch 4.07+afl
+  - opam pin add -n .
+  - opam depext cstruct-async cstruct-lwt cstruct-unix cstruct ppx_cstruct
+  - opam install -y .
+  - opam install -y crowbar fmt
+  - opam exec -- dune build fuzz/fuzz.exe
+  - mkdir out && mkdir in && echo bactrian > in/test
```
Comment (yomimono) [inline .drone.yml]: If you're using `bun` you don't need to do this anymore - `bun` should take care of it for you.

File: .drone.yml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  - opam pin add -y bun https://github.com/yomimono/ocaml-bun.git
+  - opam exec -- bun -i in -o out -- ./_build/default/fuzz/fuzz.exe
```
Comment (yomimono) [inline .drone.yml]: Here's where I'd add a `timeout` invocation.  It's done via `dune runtest` in this version, but https://github.com/yomimono/ocaml-test-stdlib/blob/primary/test/jbuild#L29 has the basic shape of how this was done in earlier `bun` invocations.

File: .drone.yml (hunk 1 above)
Comment (avsm) [reply .drone.yml]: i cant think of a pun involving bun but i pushed the change for fun and now i have to run

**Comment by talex5** [comment]
It says:
```
# Run eval $(opam env) to update the current shell environment
+ dune build @fuzz
/usr/drone/bin/init: 50: /usr/drone/bin/init: dune: not found 
```
https://cloud.drone.io/mirage/ocaml-cstruct/13/1/2

**Comment by avsm** [comment]
Hm, this is now passing but the fuzzer appears to terminate too soon despite a 30 minute timeout. https://cloud.drone.io/mirage/ocaml-cstruct/17/1/2  -- any ideas whats up here?

**Comment by talex5** [comment]
No idea. The summary is certainly disappointing:

```
Summary stats
=============

       Fuzzers alive : 48
      Total run time : 0 days, 0 hours
         Total execs : 0 million
    Cumulative speed : 0 execs/sec
       Pending paths : 48 faves, 48 total
  Pending per fuzzer : 1 faves, 1 total (on average)
       Crashes found : 0 locally unique

All fuzzers have terminated.
```

**Comment by yomimono** [comment]
@kayceesrk reported a similar bug - https://github.com/yomimono/ocaml-bun/issues/6 . I'm trying to figure out what process-handling mistake I'm making in `bun` that leads to this.

**Comment by avsm** [comment]
Hm, there have also been multiple `spawn` releases since you first wrote bun. Could be something in between like that...

**Comment by talex5** [comment]
I'm not sure I understand how it's supposed to work. From a quick look, it seems that `mon` is a signal-handler that calls `Unix.pause ()` to wait for the next signal, while the original program is also waiting on its own `Unix.pause ()`. How does that work?

The main application finishes with:

```ocaml
    Printf.printf "Fuzzers launched.  Waiting %d seconds for the first status update...\n%!" delay;
    ignore @@ Unix.alarm delay;
    Ok (Unix.pause ())
```

Does that mean that if any signal handler ever returns then the program exits?

**Comment by yomimono** [comment]
> I'm not sure I understand how it's supposed to work.

To be honest I don't really remember what was going on here. I'm working on something else this afternoon but I'll try and figure out (1) the original intent (2) whether it ever worked outside of the testing scenario where it demonstrably worked for some test code.

**Comment by talex5** [comment]
@yomimono : if you're busy with other things at the moment, I can take a look at the bun process logic.

**Comment by avsm** [comment]
And now it gets a lot of fuzzing done, hurrah! @yomimono @talex5

```
Fuzzers alive : 48
13296	Total run time : 1 days, 15 hours
13297	Total execs : 230 million
13298	Cumulative speed : 76960 execs/sec
13299	Pending paths : 0 faves, 1 total
13300	Pending per fuzzer : 0 faves, 0 total (on average)
13301	Crashes found : 0 locally unique
13302	
13303	
```

**Pull request merged**